src/dscanner/analysis/useless_initializer.d

```diff
@@ -26,301 +19,242 @@ Limitations:
  * Check that detects the initializers that are
  * not different from the implcit initializer.
  */
-final class UselessInitializerChecker : BaseAnalyzer
+// TODO: Fix NoLint
+extern (C++) class UselessInitializerChecker(AST) : BaseAnalyzerDmd
 {
-	alias visit = BaseAnalyzer.visit;
-
+	alias visit = BaseAnalyzerDmd.visit;
 	mixin AnalyzerInfo!"useless_initializer";
 
-private:
-
-	enum key = "dscanner.useless-initializer";
+	private enum KEY = "dscanner.useless-initializer";
+	private enum MSG = "Variable '%s' initializer is useless because it does not differ from the default value";
 
-	version(unittest)
+	private struct StructInfo
 	{
-		enum msg = "X";
+		string name;
+		bool shouldErrorOnInit;
+		bool isBeingVisited;
 	}
-	else
-	{
-		enum msg = "Variable `%s` initializer is useless because it does not differ from the default value";
-	}
-
-	static immutable intDefs = ["0", "0L", "0UL", "0uL", "0U", "0x0", "0b0"];
 
-	HashMap!(string, bool) _structCanBeInit;
-	DynamicArray!(string) _structStack;
-	DynamicArray!(bool) _inStruct;
-	DynamicArray!(bool) _atDisabled;
-	bool _inTest;
+	private StructInfo[string] visitedStructs;
+	private string[] structStack;
+	private bool inTest;
 
-public:
-
-	///
-	this(BaseAnalyzerArguments args)
+	extern (D) this(string fileName, bool skipTests = false)
 	{
-		super(args);
-		_inStruct.insert(false);
+		super(fileName, skipTests);
 	}
 
-	override void visit(const(Unittest) test)
+	override void visit(AST.UnitTestDeclaration unitTestDecl)
 	{
 		if (skipTests)
 			return;
-		_inTest = true;
-		test.accept(this);
-		_inTest = false;
+
+		inTest = true;
+		super.visit(unitTestDecl);
+		inTest = false;
 	}
 
-	override void visit(const(StructDeclaration) decl)
+	override void visit(AST.StructDeclaration structDecl)
 	{
-		if (_inTest)
+		if (inTest || structDecl.ident is null)
 			return;
 
-		assert(_inStruct.length > 1);
+		string structName = cast(string) structDecl.ident.toString();
+		if (isNestedStruct())
+			structName = structStack[$ - 1] ~ "." ~ structName;
```

Comment:
Why are you constructing the qualified name here? if you call "toPrettyChars" on a StructDeclaration it will give you the fully qualified name which is a good unique identifier.

Reply:
Hmm... on a second thought, it might be possible that toPrettyChars doesn't work correctly here if you are not performing semantic. Please check whether it works or not, because if it does it will save you some pain.

Reply:
Only nested structs are counted. Wouldn't `toPrettyChars` also print the encapsulating class, if present?